next.mdx.shiki.mjs

```diff
@@ -19,8 +19,150 @@ const memoizedShikiji = await getHighlighterCore({
+}
+
+/**
+ * Retrieves a list indicating the starting, and ending indexes of sequential
+ * code elements.
+ *
+ * @param {Node} tree - The current MDX resolved content.
+ *
+ * @return {{start: number, end: number}[]} - The list containing every range of
+ * sequential code elements.
+ */
+function getCodeTabsRange(tree) {
+  const rangeMap = {};
+  let start = null;
+
+  visit(tree, 'element', (node, index, parent) => {
+    // Adding 2 since there is one text node between every element
+    const next = index + 2;
+
+    if (
+      isCodeBlock(node) &&
+      'children' in parent &&
+      parent.children !== undefined &&
+      isCodeBlock(parent.children[next])
+    ) {
+      start ??= index;
+      rangeMap[start] = next;
+
+      // Prevent visiting the code block children
+      return SKIP;
+    }
+
+    // End of sequential code elements, reset the start for the next range
+    start = null;
+  });
+
+  return Object.entries(rangeMap).map(([start, end]) => ({
+    start: Number(start),
+    end: Number(end),
+  }));
+}
+
 export default function rehypeShikiji() {
   return async function (tree) {
+    // Retrieve all sequential code boxes to transform
+    const ranges = getCodeTabsRange(tree);
+
+    if (ranges.length > 0) {
+      // Make a mutable clone without reference
+      const children = [...tree.children];
+
+      for (const range of ranges) {
+        // Simple tree containing the sequential code boxes among text nodes
+        const slicedTree = {
+          type: 'root',
+          children: tree.children.slice(range.start, range.end + 1),
+        };
+
+        const languages = [];
+        const displayNames = [];
+        const codeTabsChildren = [];
+
+        visit(slicedTree, 'element', node => {
+          const codeElement = node.children[0];
+
+          const displayName = getMetaParameter(
+            codeElement.data?.meta,
+            'displayName'
+          );
+          const className = codeElement.properties.className.join(' ');
+          const matches = className.match(/language-(?<language>.*)/);
+
+          languages.push(matches.groups.language);
+          displayNames.push(displayName?.replaceAll('|', '') ?? '');
+          codeTabsChildren.push(node);
+
+          // Prevent visiting the code block children
+          return SKIP;
+        });
+
+        // Each iteration reduces the `children` length, so it needs to be
+        // accounted in the following operations
+        const lengthOffset = tree.children.length - children.length;
+        const compensatedRange = {
+          start: range.start - lengthOffset,
+          end: range.end - lengthOffset,
+        };
+        const deleteCount = compensatedRange.end - compensatedRange.start + 1;
+
+        // Replace the sequential code boxes with a code tabs element
+        children.splice(compensatedRange.start, deleteCount, {
+          type: 'element',
+          tagName: 'CodeTabs',
```

Comment:
You're making an implicit convertion to `CodeTabs`? Is this the best path here?

This pretty much creates an invisible requirement of always providing the CodeTabs component. I'm fine with it, but I wonder how... performatic this is

Reply:
I thought it was the only path, but I've managed it to work with the `pre` element. So the CodeTabs won't be needed in the mdx components map.